### DISCUSSION TITLE
Compute the max first for tree decoding

### DISCUSSION THREAD

**Discussion by vasushyam** [lucidrains/ring-attention-pytorch#19] [Ideas]
Hi, I'm one of the authors of arXiv:2408.04093. I came to know from a mutual contact that you rightly pointed out that the version of Algorithm 3 we have in v2 of the paper is not strictly overflow safe because of how the maxes are handled. Here's a reworked version that should just be stable:
<img width="565" alt="Screenshot 2024-08-14 at 2 18 55 PM" src="https://github.com/user-attachments/assets/48fc2ed4-de5a-436e-96be-39817edcbac5">

Thank you for pointing this out! and great work with the rapid implementation.

**Comment by lucidrains** [comment]
@vasushyam hey Vasudev! yes that does look good

thank you for uncovering the neat algorithm!
